경북대 BE_심규민 5주차 과제 (1단계, 2단계)

sim-mer opened the pull request:
### 궁금한 점
- 현재는 프론트가 존재하지 않기에 브라우저에서 https://kauth.kakao.com/oauth/authorize로 요청하고 로그인 화면을 얻어 그에 대한 로그인 성공으로 얻은 코드를 가져오는데 프론트가 생기고 로그인 요청이 프론트에서 들어온다면 현재 백엔드 서버에서 https://kauth.kakao.com/oauth/authorize로 요청하나요?  그렇다면 그에 대한 로그인 화면은 프론트로 어떻게 전달하는지 궁금합니다!
- 현재 코드에서 RestClient를 각 OAuthApiClient의 구현체마다 create()로 생성해서 사용하는데 전체 구현체가 모두 같은 RestClient 인스턴스를 사용하는게 더 좋은지 궁금합니다!

sim-mer committed:
chore(application.yml): kakao oauth 관련 설정 값들 추가

sim-mer committed:
feat(oauth): 임시로 redirect url에서 code값을 가져오기 위한 컨트롤러 작성

sim-mer committed:
feat(oauth): KaKao Token Api 요청을 보내기 위한 Param 작성

sim-mer committed:
feat(oauth): Kakao Api 요청 Client 작성

sim-mer committed:
feat(oauth): Kakao Access Token 받아오는 로직 작성

sim-mer committed:
test(oauth): KaKao Oauth 테스트 코드 작성

leaf-upper commented on src/main/java/gift/oauth/business/dto/KakaoParam.java:
```diff
@@ -0,0 +1,34 @@
+    private final String code;
+
+    public KakaoParam(Environment env, String code) {
+        this.grantType = env.getProperty("spring.oauth.kakao.grant_type");
```
이 3가지 값은 프로퍼티 바인딩할 클래스를 만들어서 사용해보는건 어떨까요?

sim-mer replied:
만들었습니다!

leaf-upper commented on src/main/java/gift/oauth/business/client/OAuthApiClient.java:
```diff
@@ -0,0 +1,9 @@
+import gift.oauth.business.dto.OAuthParam;
+import gift.oauth.business.dto.OAuthProvider;
+
+public interface OAuthApiClient {
```
👍 OAuth Protocol을 지원해주는곳이 많으니 이렇게 인터페이스로 빼주는건 좋은것 같아요~

sim-mer replied:
감사합니다!

leaf-upper commented on src/main/java/gift/oauth/business/client/KakaoApiClient.java:
```diff
@@ -0,0 +1,46 @@
+    private final RestClient restClient;
+
+    public KakaoApiClient() {
+        this.restClient = RestClient.create();
```
**현재 코드에서 RestClient를 각 OAuthApiClient의 구현체마다 create()로 생성해서 사용하는데 전체 구현체가 모두 같은 RestClient 인스턴스를 사용하는게 더 좋은지 궁금합니다!**

저는 개인적으로 ApiClient마다 RestClient를 별도로 생성해서 사용하는걸 선호하는데요~ 
RestClient마다 Connection Pool 및 Socket Connection Timeout, Read Timeout 등등 값을 설정해서 사용할 수 있어요.

여기에서도 비슷하게 Kakao Auth Server로 API 요청을 했을때 Connection Timeout 및 Read Timeout을 설정해보는건 어떨까요?

sim-mer replied:
시간과 Connection Pool은 설정했으나 적절한 시간과 개수를 설정하는 방법이나 팁이 궁금합니다!

leaf-upper replied:
이건 정말 API 별로 설정이 달라질 수 있을것 같고, 어떤 정해진 방법이나 팁이 있지는 않은데요~ 
저 같은 경우 처음에는 값을 보수적으로 잡았다가 이후 애플리케이션을 모니터링해서 적절한값을 찾아나가고 있습니다~

leaf-upper commented on src/main/java/gift/oauth/business/client/KakaoApiClient.java:
```diff
@@ -0,0 +1,46 @@
+            )
+            .body(OauthToken.Kakao.class);
+
+        System.out.println(result);
```
stdout 표준 입출력보다는, Logging Framework를 사용해서 로그를 남겨주시면 더 좋을것 같아요~

sim-mer replied:
넵 반영했습니다!

leaf-upper commented on src/main/java/gift/oauth/presentation/restcontroller/OauthController.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @GetMapping
+    public void kakao(@RequestParam String code) {
```
메서드 이름은 동사로 시작하면 좋을것 같아요~

sim-mer replied:
넵 수정했습니다!

leaf-upper commented on src/main/resources/application.yml:
```diff
@@ -26,3 +26,9 @@ spring:
+  oauth:
+    kakao:
+      client_id: ${kakao_client_id}
+      redirect_uri: ${kakao_redirect_uri}
```
kakao client id는 보안정보이니, 환경변수 등에서 바인딩 받는건 좋은것 같아요~
redirect uri 또한 이렇게 바인딩 받을 필요가 있을까요?

sim-mer replied:
굳이 애플리케이션 서버의 주소를 외부에 공개할 필요가 없다고 생각해서 이렇게 설정했습니다!

leaf-upper commented on src/main/java/gift/oauth/presentation/restcontroller/OauthController.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class OauthController {
```
**현재는 프론트가 존재하지 않기에 브라우저에서 [https://kauth.kakao.com/oauth/authorize로](https://kauth.kakao.com/oauth/authorize%EB%A1%9C) 요청하고 로그인 화면을 얻어 그에 대한 로그인 성공으로 얻은 코드를 가져오는데 프론트가 생기고 로그인 요청이 프론트에서 들어온다면 현재 백엔드 서버에서 [https://kauth.kakao.com/oauth/authorize로](https://kauth.kakao.com/oauth/authorize%EB%A1%9C) 요청하나요? 그렇다면 그에 대한 로그인 화면은 프론트로 어떻게 전달하는지 궁금합니다!**

이 질문과 이어지는데요~
사용자가 로그인을 요청할경우 kauth.kakao.com으로 302 redirect 응답을 주시면 될것 같아요. Flow는 아래와 같은데요,
1. 사용자가 애플리케이션 서버로 로그인 페이지 요청
2. 애플리케이션 서버에서 kauth 페이지로 사용자를 302 redirect
3. 사용자는 해당 페이지에서 로그인하고, 카카오에서는 로그인이 성공하면 해당 사용자를 다시 애플리케이션으로 redirect (이때 code를 줌.)
4. 애플리케이션 서버에서는 이 code를 기반으로 다시 access token 요청, 사용자 정보 요청 등등 진행
위 Flow대로 진행하면 될것 같아요~

sim-mer replied:
최종적으로 회원가입 or 로그인 처리 후 애플리케이션 서버에서 사용하는 Jwt 토큰을 반환하도록 변경했습니다!

leaf-upper reviewed (changes requested):
OAuth에서 Code를 받아서 AccessToken을 받는 부분은 진행되었지만, 
코멘트 드린 Flow대로 한번 로그인 구현에 도전해보면 좋을것 같아요~!
👍

sim-mer committed:
refactor(oauth): Kakao 관련 상수들 바인딩 클래로 받아오도록 변경

sim-mer committed:
refactor(oauth): 컨트롤러 로그인 메서드명 동사형으로 변경

sim-mer committed:
refactor(oauth): Logger로 확인하도록 변경

sim-mer committed:
chore(build.gradle): apache HttpClient5 설정을 위한 의존성 추가

sim-mer committed:
feat(oauth)
- RestClient Timeout, Connection Pool 설정 추가
- 로깅을 위한 InputStream -> String 유틸 메서드 추가

sim-mer committed:
refactor(util): 유틸 메서드 패키지 이동

sim-mer committed:
refactor(): logger final로 변경

sim-mer committed:
chore(application.yml): code를 얻는 url과 DB에 password를 저장하기 위한 secret 문자열 추가

sim-mer committed:
feat(util): 암호화 메서드 util에 추가

sim-mer committed:
feat(oauth): ApiClient에 이메일 받아오는 메서드 추가

sim-mer committed:
feat(oauth): 이메일 정보 요청을 위한 body, 암호화를 위한 secret Param에 추가

sim-mer committed:
refactor(oauth): 설정클래스 record로 변경, code를 받아오는 url, secret 추가

sim-mer committed:
feat(member): 주어진 이메일이 존재 여부 확인 메서드 추가

sim-mer committed:
refactor(oauth)
- 로그인 화면 리다이렉트 메서드 추가
- 코드를 받아오고 회원 가입 or 로그인 처리 후 토큰 반환하도록 변경

sim-mer committed:
feat(oauth): 카카오 API를 사용해 회원 가입 or 로그인 처리 후 토큰 반환하는 기능 구현

sim-mer commented on src/main/java/gift/oauth/presentation/restcontroller/OauthController.java:
```diff
@@ -19,9 +25,25 @@ public OauthController(OAuthService oauthService, KakaoConfig kakaoConfig) {
 
     @GetMapping
-    public void signInKakao(@RequestParam String code) {
+    public ResponseEntity<JwtToken> getKakaoCode(@RequestParam String code) {
```
여기서 리다이렉트 된 url로 코드를 받아와 Jwt토큰을 리턴하면 왜 카카오 서버로 Jwt토큰이 전달 되는게 아니라 사용자에게 전달되는지 궁금합니다!

leaf-upper replied:
카카오서버가 사용자를 리다이렉트해서 규민님 애플리케이션 서버로 요청을 보낸것이기에, 사용자에게 JWT 토큰이 넘어갈것 같아요~

sim-mer committed:
refactor(option): 옵션 수량 차감 메서드에 예외 처리 추가

sim-mer committed:
refactor(wishlist): 위시리스트 삭제 메서드에 예외 처리 추가

sim-mer committed:
feat(order): 주문 생성 기능 구현
- 주문한 만큼 상품 옵션 수량 차감
- 주문한 상품이 위시리스트에 존재할 경우 위시리스트에서 삭제
- 주문 Entity 생성 후 DB에 저장

sim-mer committed:
refactor(): OAuthProvider 패키지 이동

sim-mer committed:
refactor(member): Member Business Layer Dto refactor

sim-mer committed:
refactor(oauth): vendor 회원가입, 로그인 과정 변경
- accessToken, refreshToken, vendor를 Member Table에 저장
- password 관련 로직 제거
- MemberService에 vendor 관련 메서드 추가
- OAuthApiClient에 accessToken만 받아오는 것이 아니라 refreshToken도 같이 받아오도록 변경

sim-mer committed:
feat(order): 카카오 메시지 전송 기능 구현

sim-mer commented:
시간 관계상 2단계도 여기 PR에서 진행하겠습니다!

sim-mer committed:
refactor(order): 카카오 메시지 전송 작업 비동기로 변경

sim-mer committed:
refactor(order): 위시리스트 삭제 작업 커스텀 예외일 때만 무시하도록 변경

sim-mer committed:
refactor(order): 카카오 메시지 전송 기능 vendor 체크 로직 추가

leaf-upper commented on src/main/java/gift/global/util/EncryptionUtils.java:
```diff
@@ -0,0 +1,18 @@
+import javax.crypto.Cipher;
+import javax.crypto.spec.SecretKeySpec;
+
+public class EncryptionUtils {
```
Utils 클래스는 
1. final 키워드를 붙여서 상속이 불가능하게 끔 해주시면 좋을것 같아요
2. 생성이 불가능하게 생성자도 private으로 선언해주시면 좋을것 같아요~

sim-mer replied:
EncryptionUtils는 원래 삭제하려했는데 실수로 안했네요 삭제하겠습니다!

leaf-upper commented on src/main/java/gift/oauth/business/client/KakaoApiClient.java:
```diff
@@ -0,0 +1,142 @@
+            .build();
+    }
+
+    private ClientHttpRequestFactory getClientHttpRequestFactory() {
```
ClientHttpRequestFactory를 생성자에서 주입받을수있지않을까요?

sim-mer replied:
넵 RestClient 설정클래스를 따로 만들고 빈으로 주입받도록 하였습니다!

leaf-upper commented on src/main/java/gift/oauth/business/dto/OauthInfo.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.ZonedDateTime;
+
+public class OauthInfo {
```
OAuth와 Oauth가 혼재되어 사용되는것 같아요. 어떤방식이든 한가지로 통일해주시면 좋을것 같아요~

sim-mer replied:
통일했습니다!

leaf-upper commented on src/main/java/gift/global/util/StringUtils.java:
```diff
@@ -0,0 +1,11 @@
+import java.io.InputStream;
+import java.util.Scanner;
+
+public class StringUtils {
```
여기도 위에서 말씀드린것처럼 작업해주시면 좋을것 같아요~

sim-mer replied:
넵 알겠습니다!

leaf-upper commented on src/main/java/gift/oauth/business/client/KakaoApiClient.java:
```diff
@@ -0,0 +1,142 @@
+import org.springframework.web.client.RestClient;
+
+@Component
+public class KakaoApiClient implements OAuthApiClient {
```
OAuth 인증 API와 메세징 API는 서로 분리해주시면 좋을것 같아요.

sim-mer replied:
분리했습니다!

leaf-upper commented on src/main/java/gift/product/business/service/OrderService.java:
```diff
@@ -0,0 +1,67 @@
+    private final ProductService productService;
+    private final WishlistService wishlistService;
+    private final OrderRepository orderRepository;
+    private final ProductRepository productRepository;
```
ProductRepositor대신 ProductService에 메서드를 위임해서 사용해보는건 어떨까요?

sim-mer replied:
ProductService, MemberService에 위임했습니다!

leaf-upper commented on src/main/java/gift/product/persistence/entity/Order.java:
```diff
@@ -0,0 +1,54 @@
+    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "product_id")
+    private Product product;
```
이 코멘트는 생각만 해보면 좋을거 같아요.
- 주문은 **"특정 시간대의"** 주문 이기 때문에, 상품의 데이터도 **"주문 당시의 상품"** 이어야해요. 현재 코드에서는  상품데이터가 변경되면, 연관된 주문데이터가 변경된것과 같아요.
- 그렇기 때문에 "주문상품"과 "상품"이 가지는 의미가 약간 다를 수 있어요. 주문 상품은 "주문 당시의 상품 데이터" 여야하고, 상품은 "현재 시간의 상품"이어야 해요.

sim-mer replied:
넵 알겠습니다!

leaf-upper reviewed (approved):
👍 고생하셨습니다~ 몇가지 코멘트 드린부분은 생각해보시고 3단계 진행할때 반영해주시면 좋을것 같아요 
2단계까지 API를 연동해보셨는데 많이 어려우셨을것 같네요. 정말 고생많으셨습니다 👍
